### PR TITLE
CI: add windows ubsan and clang workflows

### PR DESCRIPTION
**Pull request by mizvekov**

### CODE REVIEW

**Comment by mizvekov** [comment]
FYI this is why the clang-windows job is failing: https://github.com/libuv/libuv/issues/4139

**Comment by vtjnash** [comment]
Does it work with https://github.com/libuv/libuv/pull/4185 now?

File: .github/workflows/sanitizer.yml [hunk 1]
```diff
@@ -60,3 +60,18 @@ jobs:
...
+    runs-on: windows-2022
+    steps:
+      - uses: actions/checkout@v2
+      - name: Envinfo
+        run: npx envinfo
+
+      - name: UBSAN Build
+        run: |
+          mkdir build-ubsan
+          cmake -B build-ubsan -G Ninja -DBUILD_TESTING=ON -DUBSAN=ON -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_COMPILER=clang
```
Comment (vtjnash) [inline .github/workflows/sanitizer.yml:73]:
```suggestion
          cmake -B build-ubsan -G "Visual Studio 17 2022" -DBUILD_TESTING=ON -DUBSAN=ON -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_COMPILER=clang
```

Why is this added as both a sanitizers and CI-win workload? Do we just need one?

**Comment by mizvekov** [comment]
@vtjnash I just rebased, but this was no-test. There is something different with the MSVC CI now, compared to last month, where before 'Ninja' tool was available, now is not.
So we are failing before we even try to generate the build system, which does not test your changes.

File: .github/workflows/sanitizer.yml (hunk 1 above)
Comment (mizvekov) [reply .github/workflows/sanitizer.yml:73]: One config is testing pure Visual Studio generator, plus MSVC's CL, and the other is testing Ninja generator, with clang as compiler.

The build failure your MR is trying to fix was affecting Ninja generator only, so this suggested change would hide the problem.

Note that using Ninja generator with Visual Studio is pretty common when using it's cmake integration, as it is fully supported and produces much faster builds.

Comment (vtjnash) [reply .github/workflows/sanitizer.yml:73]: Looks like it must be installed explicitly now:
https://github.com/actions/runner-images/issues/8343#issuecomment-1727810519

**Comment by mizvekov** [comment]
@vtjnash Looks good now, we generate a valid ninja.build, which gets us farther than before.

Now it's failing yet, for a different reason, not the one I intended originally. It's the same problem as this: https://github.com/microsoft/STL/issues/3568
This issue is easy to work around though, there is a clang flag which makes the sanitizer CodeGen just emit traps instead of calls to the ubsan runtime. I will make this change soon enough, it's getting late here.

After that is working, I would expect this CI job to fail once again, but this time due to UB when running the windows tests, which is what I intended to demonstrate initially.

File: test/test-tcp-close-reset.c [hunk 2]
```diff
@@ -81,7 +81,7 @@ static void do_close(uv_tcp_t* handle) {
     ASSERT_OK(r);
 #ifdef _WIN32
     ASSERT_PTR_NE(fd, INVALID_FD);
-    ASSERT_OK(shutdown(fd, SD_BOTH));
+    ASSERT_OK(shutdown((uintptr_t)fd, SD_BOTH));
```
Comment (vtjnash) [inline test/test-tcp-close-reset.c:84]:
```suggestion
    ASSERT_OK(shutdown((SOCKET)fd, SD_BOTH));
```

**Review by vtjnash — APPROVED** [review]

**Comment by mizvekov** [comment]
Alright, thanks for the review!
You can go ahead and merge this.

**Pull request merged**